Authentication/SignUp/View/SignupController.swift

```diff
@@ -187,12 +189,45 @@ private extension SignupController {
     private func setTextfieldOrder() {
         signupView.usernameTextField?.nextTextField = signupView.emailTextField
         signupView.emailTextField.nextTextField = signupView.passwordTextField
-        signupView.passwordTextField.nextTextField = signupView.passwordConfirmTextField ?? signupView.usernameTextField ?? signupView.emailTextField
-        signupView.passwordConfirmTextField?.nextTextField = signupView.usernameTextField ?? signupView.emailTextField
+        signupView.passwordTextField.nextTextField = passwordNextTextfield
+        signupView.passwordConfirmTextField?.nextTextField = _viewModel.usernameEnabled ? signupView.usernameTextField : signupView.emailTextField
+        lastTextField.returnKeyType = .Go
+    }
+    
+    private var passwordNextTextfield: UITextField? {
+        if _viewModel.passwordConfirmationEnabled {
+            return signupView.passwordConfirmTextField
+        } else {
+            if _viewModel.usernameEnabled {
+                return signupView.usernameTextField
+            } else {
+                return signupView.emailTextField
+            }
+        }
     }
     
     private var lastTextField: UITextField {
-        return signupView.passwordConfirmTextField ?? signupView.passwordTextField
+        return _viewModel.passwordConfirmationEnabled ? signupView.passwordConfirmTextField! : signupView.passwordTextField
+    }
+    
+    private func hideUnselectedTextfields() {
+        if !_viewModel.usernameEnabled {
+            signupView.hideUsernameElements()
+        }
+        if !_viewModel.passwordConfirmationEnabled {
+            signupView.hidePasswordConfirmElements()
+        }
+    }
+    
+    private func checkTextFieldsSelection() {
+        if _viewModel.usernameEnabled && signupView.usernameTextField == .None {
+            NSLog("The signup view and view model are not compatible. The view model will try to validate content of username textfield which doesn't exist in the view.")
+            fatalError("The signup view and view model are not compatible.")
```

Comment:
But it is specified in the Log. Is the log message okay? I could just copy that here.